conference/cart.py

```diff
@@ -0,0 +1,297 @@
+
+        discount_code, fares_info = extract_order_parameters_from_request(
+            request.POST
+        )
+        context['fares_info'] = fares_info
+
+        try:
+            calculation, coupon = calculate_order_price_including_discount(
+                for_user=request.user,
+                for_date=timezone.now().date(),
+                fares_info=fares_info,
+                discount_code=discount_code,
+            )
+        except FareIsNotAvailable:
+            return HttpResponse("Fare is not available", status=403)
+
+        if CartActions.apply_discount_code in request.POST:
+            context['calculation'] = calculation
+            context['discount_code'] = discount_code or 'No discount code'
+
+        if CartActions.buy_tickets in request.POST:
+            order = create_order(
+                for_user=request.user,
+                for_date=timezone.now().date(),
+                fares_info=fares_info,
+                calculation=calculation,
+                coupon=coupon,
+            )
+            return redirect(
+                "cart:step3_add_billing_info", order_uuid=order.uuid
+            )
+
+    return TemplateResponse(
+        request, "ep19/bs/cart/step_2_pick_tickets.html", context
+    )
+
+
+@login_required
+@staff_member_required  # TEMPORARY for testing
+def cart_step3_add_billing_info(request, order_uuid):
+
+    order = get_object_or_404(Order, uuid=order_uuid)
+
+    if is_business_order(order):
+        billing_form = BusinessBillingForm
+    else:
+        # TODO: should we have a student billing form?
+        billing_form = PersonalBillingForm
+
+    form = billing_form(instance=order)
+
+    if request.method == 'POST':
+        form = billing_form(data=request.POST, instance=order)
+
+        if form.is_valid():
+            form.save()
+            # TODO(artcz)
+            # if form billing notes then create custom helpdesk case?
+
+            return redirect("cart:step4_payment", order_uuid=order.uuid)
+
+    return TemplateResponse(
+        request, "ep19/bs/cart/step_3_add_billing_info.html", {'form': form}
+    )
+
+
+@login_required
+@staff_member_required  # TEMPORARY for testing
+def cart_step4_payment(request, order_uuid):
+    order = get_object_or_404(Order, uuid=order_uuid)
+    total_for_stripe = int(order.total() * 100)
+    payments = order.stripepayment_set.all().order_by('created')
+    # TODO: pay attention if order is paid or not, and if paid don't render the
+    # payment button in the template
+
+    if request.method == 'POST':
+        stripe_payment = StripePayment.objects.create(
+            order=order,
+            user=request.user,
+            uuid=str(uuid.uuid4()),
+            amount=order.total(),
+            token=request.POST.get('stripeToken'),
+            token_type=request.POST.get('stripeTokenType'),
+            email=request.POST.get('stripeEmail'),
+        )
+        try:
+            with transaction.atomic():
```

Comment:
Don't think this block should be atomic - when testing and having the invoice generation to fail, the payment is recorded without a status, even though stripe managed to charge the account, which is misleading. Splitting it into two atomic blocks would make more sense:
```
with transaction.atomic():
    charge_for_payment(stripe_payment)
    order.payment_date = date.today()
    order.save()
with transaction.atomic():
    create_invoices_for_order(order)
    # TODO send_confirmation_email()```